File: java_solutions/331.verify-preorder-serialization-of-a-binary-tree.java
/*
 * @lc app=leetcode id=331 lang=java
 *
 * [331] Verify Preorder Serialization of a Binary Tree
 */

class Solution {
    public boolean isValidSerialization(String preorder) {
        return java.util.Arrays.stream(preorder.split(",")).reduce(1, (diff, node) -> diff <= 0 ? -1 : diff - 1 + (node.equals("#") ? 0 : 2), (a, b) -> a) == 0;
    }
}

File: kotlin_solutions/331.verify-preorder-serialization-of-a-binary-tree.kt
/*
 * @lc app=leetcode id=331 lang=kotlin
 *
 * [331] Verify Preorder Serialization of a Binary Tree
 */

class Solution {fun isValidSerialization(preorder: String) = preorder.split(",").fold(1 to true) { (slots, valid), str -> if (!valid || slots <= 0) (slots to false) else ((slots - 1 + if (str != "#") 2 else 0) to true) }.let { (slots, valid) -> valid && slots == 0 }}

